### PR TITLE
Chore: Add download reachability checks to Document and Image viewers

### PR DESCRIPTION
**Pull request by jeremypress**
This also will add the check to any generic download request, such as prefetching or preloading. 

Tests on the way, other viewers will be in a separate PR.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @jeremypress has signed the CLA. Thanks for the pull request!

File: src/i18n/en-US.properties [hunk 1]
```diff
@@ -172,6 +172,8 @@ notification_button_default_text=Okay
 notification_annotation_point_mode=Click anywhere to add a comment to the document
 # Notification message shown when user enters drawing annotation mode
 notification_annotation_draw_mode=Press down and drag the pointer to draw on the document
+# Notification message shown when the user has a degraded preview experience due to blocked download hosts
+notification_degraded_preview=Your preview experience is degraded!
```
Comment (jeremypress) [inline src/i18n/en-US.properties]: Not the final text

File: src/lib/viewers/image/ImageBaseViewer.js [hunk 2]
```diff
@@ -311,23 +310,27 @@ class ImageBaseViewer extends BaseViewer {
...
-    errorHandler(err) {
-        /* eslint-disable no-console */
-        console.error(err);
-        /* eslint-enable no-console */
-
-        // Display a generic error message but log the real one
-        const error = err;
-        if (err instanceof Error) {
-            error.displayMessage = __('error_refresh');
+    handleDownloadError(err, imgUrl) {
```
Comment (JustinHoldstock) [inline src/lib/viewers/image/ImageBaseViewer.js]: This is the same as `ImageBaseViewer.handleDownloadError()`. Can you make this a function in the parent class?

Comment (jeremypress) [reply src/lib/viewers/image/ImageBaseViewer.js]: Do you mean in docBase vs. imageBase? The doc version of this function throws a different error (error_document vs. error_refresh, I need to fix that), and other viewers will want custom behavior here as well. See how the dash-viewer handles download errors currently.

File: src/lib/util.js [hunk 3]
```diff
@@ -111,6 +116,63 @@ function createDownloadIframe() {
...
 }
 
+/**
+ * Checks if the url is a download host, but not the default download host.
+ *
+ * @public
+ * @param {string} url - The URL to check
+ * @return {boolean} - HTTP response
+ */
+export function isNonDefaultDownloadHost(url) {
```
Comment (pramodsum) [inline src/lib/util.js]: Could we use something like `isCustomDefaultDownloadHost`? Feels weird to have isNon.... in the method name

File: src/lib/util.js [hunk 4]
```diff
@@ -111,6 +116,63 @@ function createDownloadIframe() {
...
+}
+
+/**
+ * Stores the host in an array via localstorage so that we don't show a notification for it again
+ *
+ * @public
+ * @param {string} downloadHost - Download URL host
+ * @return {void}
+ */
+export function setDownloadHostNotificationShown(downloadHost) {
```
Comment (pramodsum) [inline src/lib/util.js]: `cacheDownloadHostNotificationShown`?

File: src/lib/viewers/BaseViewer.js [hunk 5]
```diff
@@ -384,13 +406,27 @@ class BaseViewer extends EventEmitter {
     }
 
     /**
-     * Handles the viewer load to potentially set up Box Annotations.
```
Comment (pramodsum) [inline src/lib/viewers/BaseViewer.js]: 👍

File: src/lib/viewers/image/ImageBaseViewer.js (hunk 2 above)
Comment (pramodsum) [reply src/lib/viewers/image/ImageBaseViewer.js]: If the only difference between this and docBase is the error message, wouldn't it make sense to use this as the default and maybe pass in an error message in each viewer? The rest of the code seems to stay the same

Comment (jeremypress) [reply src/lib/viewers/image/ImageBaseViewer.js]: Looking at it again you're right, the only significant difference are the messages, so I'll pass those in.

File: src/lib/util.js (hunk 4 above)
Comment (jeremypress) [reply src/lib/util.js]: It's a little more extreme than a cache since we are using localStorage, but I'll think some more about the name.

**Pull request merged**